File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if(!root) return;
        if(!root->left) {
            flatten(root->right);
            return;
        }
        if(!root->right) {flatten(root->left);
                          root->right=root->left;
                          root->left=NULL;
                          return;
                         }
        TreeNode* temp=root->right;
        flatten(root->left);
        if(root->left) root->right=root->left;
        root->left=NULL;
        TreeNode* f=root;
        while(f->right){
            f=f->right;
        }
        f->right=temp;
        if(f->right) flatten(f->right);
        return;
    }
};

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    prev = None
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        
        if root:
            self.flatten(root.right)
            self.flatten(root.left)
            root.left = None
            root.right = self.prev
            self.prev = root
            
            
            # temp = root.right
            # while root.right:
            #     root = root.right
            # root.right = temp
#         def traverse(node):
#             if not node:
#                 return None
#             if node.left:
#                 n = node.left
                
#                 while n.right:
#                     n = n.right
#                 n.right = node.right
#                 node.right = node.left
#                 node.left = None
#             traverse(node.right)
#         traverse(root)
        